src/lib.rs

```diff
@@ -549,48 +554,118 @@ impl<'input> Namespace<'input> {
     ///     "<e xmlns:n='http://www.w3.org'/>"
     /// ).unwrap();
     ///
-    /// assert_eq!(doc.root_element().namespaces()[0].uri(), "http://www.w3.org");
+    /// assert_eq!(doc.root_element().namespaces().nth(0).unwrap().uri(), "http://www.w3.org");
     /// ```
     #[inline]
     pub fn uri(&self) -> &str {
         self.uri.as_ref()
     }
 }
 
-struct Namespaces<'input>(Vec<Namespace<'input>>);
+#[derive(Default)]
+struct Namespaces<'input> {
+    // Deduplicated namespace values used throughout the document
+    values: Vec<Namespace<'input>>,
+    // Indices into the above in tree order as the document is parsed
+    tree: Vec<u16>,
+    // Indices into the above sorted by value used for deduplication
+    sorted: Vec<u16>,
+}
 
 impl<'input> Namespaces<'input> {
-    #[inline]
-    fn push_ns(&mut self, name: Option<&'input str>, uri: Cow<'input, str>) {
+    fn push_ns<'temp>(
+        &mut self,
+        name: Option<&'input str>,
+        uri: BorrowedText<'input, 'temp>,
+    ) -> Result<(), Error> {
         debug_assert_ne!(name, Some(""));
-        self.0.push(Namespace { name, uri });
+
+        let idx = match self.sorted.binary_search_by(|idx| {
+            let value = &self.values[*idx as usize];
+
+            (value.name, value.uri.as_ref()).cmp(&(name, uri.as_str()))
+        }) {
+            Ok(sorted_idx) => self.sorted[sorted_idx],
+            Err(sorted_idx) => {
+                if self.values.len() > core::u16::MAX as usize {
+                    return Err(Error::NamespacesLimitReached);
+                }
+                let idx = self.values.len() as u16;
+                self.values.push(Namespace {
+                    name,
+                    uri: uri.to_cow(),
+                });
+                self.sorted.insert(sorted_idx, idx);
```

Comment:
I would say both are valid strategies: Using two vectors provides better data locality and code density but makes insertion asymptotically slower as you mentioned. `BTreeSet` will likely be faster for larger numbers of unique names, but will also need more code and have worse data locality.

So I'd say we need to measure and to do that determine the target, i.e. how many unique names and how many usages of names.

Reply:
(Note that this is actually part of #80.)

Reply:
(Note that this is actually part of #80.)

Reply:
@tomjw64 A typical XML has like 1-5 namespaces. I don't think we would hit any performance issues here.

Reply:
@RazrFalcon If we really want to make this assumption, we can drop the binary search completely as a simple linear search will do and the separate `sorted` list will be unnecessary. It will mean more significant performance degradation for files with a larger number of namespaces though. (For example, that CSW response has 15 namespaces and I regularly process RDF files with a similar number of namespaces.)

Reply:
So, to maybe add some data on this:

`large.plist` sees a small slowdown when the binary search is replaced by a linear search:

```
Benchmark 1: ./stats-linear-search benches/large.plist
  Time (mean ± σ):       5.7 ms ±   0.3 ms    [User: 4.5 ms, System: 0.9 ms]
  Range (min … max):     5.0 ms …   6.7 ms    523 runs
 
Benchmark 2: ./stats-binary-search benches/large.plist
  Time (mean ± σ):       5.5 ms ±   0.4 ms    [User: 4.5 ms, System: 0.9 ms]
  Range (min … max):     4.9 ms …   6.7 ms    535 runs
 
Summary
  './stats-binary-search benches/large.plist' ran
    1.03 ± 0.09 times faster than './stats-linear-search benches/large.plist'
```

which is strange since that file does not use namespaces, so I guess that the simpler linear search leads to worse inlining decisions?

`medium.svg` which does actually have 8 namespaces also sees a slow down:

```
Benchmark 1: ./stats-linear-search benches/medium.svg
  Time (mean ± σ):       2.3 ms ±   0.1 ms    [User: 1.8 ms, System: 0.4 ms]
  Range (min … max):     2.0 ms …   3.1 ms    1190 runs
 
Benchmark 2: ./stats-binary-search benches/medium.svg
  Time (mean ± σ):       2.2 ms ±   0.1 ms    [User: 1.7 ms, System: 0.4 ms]
  Range (min … max):     2.0 ms …   3.0 ms    1430 runs
 
Summary
  './stats-binary-search benches/medium.svg' ran
    1.06 ± 0.09 times faster than './stats-linear-search benches/medium.svg'

```

My CSW response is similarly faster with binary search:

```
Benchmark 1: ./stats-linear-search /home/adam/Downloads/q.xml
  Time (mean ± σ):     484.6 ms ±   3.4 ms    [User: 451.1 ms, System: 33.2 ms]
  Range (min … max):   479.5 ms … 490.8 ms    10 runs
 
Benchmark 2: ./stats-binary-search /home/adam/Downloads/q.xml
  Time (mean ± σ):     477.2 ms ±   8.1 ms    [User: 443.2 ms, System: 33.6 ms]
  Range (min … max):   472.3 ms … 499.1 ms    10 runs
 
Summary
  './stats-binary-search /home/adam/Downloads/q.xml' ran
    1.02 ± 0.02 times faster than './stats-linear-search /home/adam/Downloads/q.xml'
```

So while it may as much due to inlining decisions as due to algorithmic choices, it seems binary search is a bit better than linear search contrary to wait I said above.

Will look into `BTreeSet` next...

Reply:
Feel free to add a new XML sample to the benchmark if needed. Well, as long as it's not copyrighted or something.

Reply:
@tomjw64 When trying to implement it, I noticed that using a `BTreeSet` is not that simple: We need a stable order of the values as we parse the document (otherwise we would need to fix up existing indices as the order changes). Hence I could only replaced `sorted(_order)` with a `BTreeMap<Namespace<'input>, u16>` with stable indexes into the `values` list.

But this would mean that all namespaces which need allocation would have to be allocated twice and this structure would have a larger cache footprint. I am pretty sure this would be worse than shuffling around a few `u16`s on insertion.

Reply:
So in summary, its seems like a flat list of `u16` and binary search is still the best option?

Reply:
> @tomjw64 A typical XML has like 1-5 namespaces. I don't think we would hit any performance issues here.

We aren't just using this technique for Namespaces, but also for ExpandedNames in this PR. We are not talking about just 5 namespaces in this case, right? Are we also unconcerned by the potential number of unique expanded names?

Reply:
> @tomjw64 When trying to implement it, I noticed that using a `BTreeSet` is not that simple: We need a stable order of the values as we parse the document (otherwise we would need to fix up existing indices as the order changes). Hence I could only replaced `sorted(_order)` with a `BTreeMap<Namespace<'input>, u16>` with stable indexes into the `values` list.
> 
> But this would mean that all namespaces which need allocation would have to be allocated twice and this structure would have a larger cache footprint. I am pretty sure this would be worse than shuffling around a few `u16`s on insertion.

I don't dispute that `tree_order` should be a `Vec`, but as for `sorted_order`, could we try  a `BTreeSet<NewType>` where `struct NewType(u32/u16)` and `NewType::cmp` indexes into `values` to make a comparison?

Reply:
@tomjw64 My bad. I thought this code applies only to namespaces.

@adamreichold Yes. Seems to be good enough.

Reply:
> Feel free to add a new XML sample to the benchmark if needed. Well, as long as it's not copyrighted or something.

Will do in #84, but I guess I will aim for something in the sub 1MB range as it will going into Git.

> We aren't just using this technique for Namespaces, but also for ExpandedNames in this PR. We are not talking about just 5 namespaces in this case, right? Are we also unconcerned by the potential number of unique expanded names?

> My bad. I thought this code applies only to namespaces.

Well the comment is attached to the place were namespaces are handled, but yes deduplication of expanded names as presented here uses the same technique (but with a `u32` instead of a `u16` index).

I basically agree that for tag and attribute names, the trade-offs are different and much more distinct names are possible. I still think that the flat sorted list is the best approach due to locality, but even with this I repeated the benchmarks against `master` after deduplication of namespaces with the change to `u16` and avoiding redundant allocations was merged and they look rather bad now as the baseline has gotten faster:

```
 name                                    master ns/iter  dedup ns/iter  diff ns/iter   diff %  speedup 
 large_roxmltree                         3,512,675       4,252,045           739,370   21.05%   x 0.83 
 medium_roxmltree                        818,021         1,048,494           230,473   28.17%   x 0.78 
 roxmltree_iter_children                 3,626           3,006                  -620  -17.10%   x 1.21 
 roxmltree_iter_descendants_expensive    121,392         120,818                -574   -0.47%   x 1.00 
 roxmltree_iter_descendants_inexpensive  39,760          37,976               -1,784   -4.49%   x 1.05 
 tiny_roxmltree                          5,211           5,931                   720   13.82%   x 0.88 
```

So I think we should _not_ do this any more.

I am sorry for all the hassle I put you guys though, but the memory savings from 6.7 GB to 5.5 GB are hard to ignore and I hoped this technique would improve things enough to make this workable but alas it does not.

> As for sorted_order, could we try a BTreeSet<NewType> where struct NewType(u32/u16) and NewType::cmp indexes into values to make a comparison?

The problem is that `BTreeSet<NewType>` requires `NewType: Ord`, i.e. implementing `fn cmp(&self, other: &Self) -> Ordering;`, so there is no place to slip in the reference to `values` which is need to resolve `NewType` into something that can be reasonably sorted. (One could try to make `NewType(u16, &[Namespace<'input>])` but that would make it redundantly large and `struct Namespaces` self-referential.)

Reply:
Also note that this slow down is unlikely to be due to the algorithmic complexity of the flat sorted list as e.g. `large.plist` has 8 unique element and attribute names whereas `fonts.conf` has 9. Only `medium.svg` has 148 names which might explain why it is the worst of the three but even 148 `u16`s are only `(148 * 2 + 63) / 64 = 5` cache lines.

It just seems that deduplicating namespace is a bit of a sweet spot as we already invest quite a bit of effort into their processing so that the extra cost of the deduplication doesn't impact things too badly while their inheritance rules means they will be referenced a lot of (`tree_order` is quite large) even though there are only a few distinct values (`values` is quite small).

Reply:
Makes sense why the BTreeSet won't work. Super thorough, thanks.